generator/src/test/java/com/scottlogic/deg/generator/outputs/targets/FileOutputTargetTests.java

```diff
@@ -0,0 +1,86 @@
+import org.junit.jupiter.api.Test;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class FileOutputTargetTests {
+    //Create Mocks
+    private Path mockFilePath = mock(Path.class);
+    private FileUtils mockFileUtils = mock(FileUtils.class);
+    private OutputFormat mockOutputFormat = mock(OutputFormat.class);
+    private Path mockParentPath = mock(Path.class);
+
+    @Test
+    public void validate_generateOutputFileIsADirectory_throwsException(){
+        //Arrange
+        when(mockFileUtils.isDirectory(any())).thenReturn(true);
+        when(mockFilePath.getParent()).thenReturn(mockParentPath);
+        FileOutputTarget outputTarget = new FileOutputTarget(mockFilePath, mockOutputFormat, false, mockFileUtils);
+
+        //Act
+        OutputTargetValidationException thrown = assertThrows(OutputTargetValidationException.class, ()->outputTarget.validate(),"Expected OutputTargetValidationException to throw, but didn't");
+
+        //Assert
+        assertEquals( "target is a directory, please use a different output filename", thrown.getMessage());
+    }
+
+    @Test
+    public void validate_generateOutputFileAlreadyExistsNoOverwrite_throwsException() {
+        //Arrange
+        when(mockFileUtils.isDirectory(mockFilePath)).thenReturn(false);
+        when(mockFileUtils.isDirectory(mockParentPath)).thenReturn(true);
+        when(mockFilePath.getParent()).thenReturn(mockParentPath);
+        when(mockFileUtils.exists(any())).thenReturn(true);
+        FileOutputTarget outputTarget = new FileOutputTarget(mockFilePath, mockOutputFormat, false, mockFileUtils);
+
+        //Act
+        OutputTargetValidationException thrown = assertThrows(OutputTargetValidationException.class, ()->outputTarget.validate(),"Expected OutputTargetValidationException to throw, but didn't");
+
+        //Assert
+        assertEquals( "file already exists, please use a different output filename or use the --replace option", thrown.getMessage());
+    }
+
+    @Test
+    public void validate_generateOutputFileAlreadyExistsOverwrite_doesntThrow() {
+        when(mockFileUtils.isDirectory(mockFilePath)).thenReturn(false);
+        when(mockFileUtils.isDirectory(mockParentPath)).thenReturn(true);
+        when(mockFilePath.getParent()).thenReturn(mockParentPath);
+        when(mockFileUtils.exists(any())).thenReturn(true);
+        FileOutputTarget outputTarget = new FileOutputTarget(mockFilePath, mockOutputFormat, true, mockFileUtils);
+
+        assertDoesNotThrow(() ->outputTarget.validate(),"Expected no exception, but one was thrown");
+    }
+
+    @Test
+    public void validate_generateOutputFileDoesntExist_doesntThrow() throws IOException {
+        mockFilePath = Paths.get("/a/b/c/tmp.out");
```

Comment:
Personally I think it can sometimes be helpful to randomly generate the mock path in situations like this, so that you know you are doing a wide-ranging test of the possible input space.

Other people however disagree with me intensely on this - on the grounds it makes the test non-deterministic.  My argument is: with a test like this as-is, you are effectively only testing that the code works in one special case.  In this case: what if you had had some early boilerplate code which was done before command-line parsing was finished and had hardcoded the output filename as `tmp.out` somewhere?  This test might not detect that the code was broken.

Reply:
I would disagree with you here Will, as I'm one of those against non-deterministic tests.

For this test, I think it would be better just to remove the `mockFilePath = Paths.get("/a/b/c/tmp.out");` line, because I think the test would still be equivalent in this case, but without the potential side effects you've mentioned.

Testing for the Path being correct should not be done in this file, I don't think.

Reply:
The test causes a NullPointer if the line is removed but works with any string provided as a path, regardless of if it is correctly formatted. What would be the easiest way to randomly generate a path?

Reply:
By just generating a random string, in that case, bearing in mind that it must be a valid path string.

`Paths.get()` gets a path from the default `Filesystem`, so you might want to be restrictive about what characters are in your random string, because strictly speaking you don't know what characters will be legal and what won't be (for example on Windows you can't have a path containing the `*` character; on Unix you generally can even though it is probably a bad idea)